contracts/example/ZoraWrapper/ZoraWrapper.sol

```diff
@@ -0,0 +1,241 @@
+ * - If auction finishes with `cancelAuction`:
+ *   - This contract gets wrapped voucher back and the bidder gets the price.
+ *   - `unwrap` can be executed by the owner of the wrapped voucher.
+ *
+ * N.B. Although Zora Auction House can send ethers, it's preffered to receive
+ * WETH instead. For that reason `recieve` is not implemented, so it automatically sends WETH.
+ */
+contract ZoraWrapper is BosonTypes, Ownable, ERC721 {
+    // Add safeTransferFrom to IERC20
+    using SafeERC20 for IERC20;
+
+    // Contract addresses
+    address private immutable voucherAddress;
+    address private immutable zoraAuctionHouseAddress;
+    address private immutable protocolAddress;
+    address private immutable wethAddress;
+
+    // Token ID for which the price is not yet known
+    uint256 private pendingTokenId;
+
+    // Mapping from token ID to price. If pendingTokenId == tokenId, this is not the final price.
+    mapping(uint256 => uint256) private price;
+
+    // Mapping to cache exchange token address, so costly call to the protocol is not needed every time.
+    mapping(uint256 => address) private cachedExchangeToken;
+
+    /**
+     * @notice Constructor
+     *
+     * @param _voucherAddress The address of the voucher that are wrapped by this contract.
+     * @param _zoraAuctionHouseAddress The address of Zora Auction House.
+     */
+    constructor(
+        address _voucherAddress,
+        address _zoraAuctionHouseAddress,
+        address _protocolAddress,
+        address _wethAddress
+    ) ERC721(getVoucherName(_voucherAddress), getVoucherSymbol(_voucherAddress)) {
+        voucherAddress = _voucherAddress;
+        zoraAuctionHouseAddress = _zoraAuctionHouseAddress;
+        protocolAddress = _protocolAddress;
+        wethAddress = _wethAddress;
+
+        // Approve Zora Auction House to transfer wrapped vouchers
+        _setApprovalForAll(address(this), _zoraAuctionHouseAddress, true);
+        _setApprovalForAll(address(this), msg.sender, true); // msg.sender is the owner of this contract and must be approved to transfer wrapped vouchers to Auction House
+    }
+
+    /**
+     * @dev Returns true if this contract implements the interface defined by
+     * `interfaceId`. See the corresponding
+     * https://eips.ethereum.org/EIPS/eip-165#how-interfaces-are-identified[EIP section]
+     * to learn more about how these ids are created.
+     */
+    function supportsInterface(bytes4 _interfaceId) public view virtual override(ERC721) returns (bool) {
+        return (_interfaceId == type(IERC721).interfaceId || _interfaceId == type(IERC165).interfaceId);
+    }
+
+    /**
+     * @notice Wraps the voucher, transfer true voucher to itself and funds to the protocol.
+     *
+     * Reverts if:
+     *  - caller is not the contract owner
+     *
+     * @param _tokenId The token id.
+     */
+    function wrap(uint256 _tokenId) external onlyOwner {
+        // should wrapping be limited to onlyOwner?
```

Comment:
I think we should allow at least _tokenId owner to call, but for that we also need to change the logic in unwrap function not to check if caller is contract owner.